actors/ethaccount/src/lib.rs

```diff
@@ -0,0 +1,150 @@
+pub mod types;
+
+use fvm_ipld_encoding::ipld_block::IpldBlock;
+use fvm_shared::address::{Payload, Protocol};
+use fvm_shared::crypto::hash::SupportedHashes::Keccak256;
+use fvm_shared::error::ExitCode;
+use fvm_shared::{MethodNum, METHOD_CONSTRUCTOR};
+use num_derive::FromPrimitive;
+
+use crate::types::AuthenticateMessageParams;
+use fil_actors_runtime::runtime::{ActorCode, Runtime};
+use fil_actors_runtime::{
+    actor_dispatch, actor_error, ActorDowncast, ActorError, AsActorError, EAM_ACTOR_ID,
+    FIRST_EXPORTED_METHOD_NUMBER, SYSTEM_ACTOR_ADDR,
+};
+
+#[cfg(feature = "fil-actor")]
+fil_actors_runtime::wasm_trampoline!(EthAccountActor);
+
+/// Ethereum Account actor methods.
+#[derive(FromPrimitive)]
+#[repr(u64)]
+pub enum Method {
+    Constructor = METHOD_CONSTRUCTOR,
+    AuthenticateMessageExported = frc42_dispatch::method_hash!("AuthenticateMessage"),
+}
+
+/// Ethereum Account actor.
+pub struct EthAccountActor;
+
+impl EthAccountActor {
+    /// Ethereum Account actor constructor.
+    /// NOTE: This method is NOT currently called from anywhere, instead the FVM just deploys EthAccounts.
+    pub fn constructor(rt: &mut impl Runtime) -> Result<(), ActorError> {
+        rt.validate_immediate_caller_is(std::iter::once(&SYSTEM_ACTOR_ADDR))?;
+
+        match rt
+            .lookup_delegated_address(rt.message().receiver().id().unwrap())
+            .map(|a| *a.payload())
+        {
+            Some(Payload::Delegated(da)) if da.namespace() == EAM_ACTOR_ID => {}
+            Some(_) => {
+                return Err(ActorError::illegal_argument(
+                    "invalid target for EthAccount creation".to_string(),
+                ));
+            }
+            None => {
+                return Err(ActorError::illegal_argument(
+                    "receiver must have a predictable address".to_string(),
+                ));
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Authenticates whether the provided signature is valid for the provided message.
+    /// Should be called with the raw bytes of a signature, NOT a serialized Signature object that includes a SignatureType.
+    /// Errors with USR_ILLEGAL_ARGUMENT if the authentication is invalid.
+    pub fn authenticate_message(
+        rt: &mut impl Runtime,
+        params: AuthenticateMessageParams,
+    ) -> Result<(), ActorError> {
+        rt.validate_immediate_caller_accept_any()?;
+        let msg_hash = rt.hash_blake2b(&params.message);
+
+        let signer_pk = rt
+            .recover_secp_public_key(
+                &msg_hash,
+                params
+                    .signature
+                    .as_slice()
+                    .try_into()
+                    .map_err(|_| actor_error!(illegal_argument; "invalid signature length"))?,
+            )
+            .map_err(|e| {
+                e.downcast_default(
```

Comment:
Please use `.context_code()` rather than introduce any new uses of `downcast_default`.